File: leetcode/450.delete-node-in-a-bst.cpp
// @leet start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
 public:
  TreeNode* deleteNode(TreeNode* root, int key) {
    // Get a pointer to the pointer to target node
    auto** toDeletePtr = &root;

    for (;;) {
      if (!*toDeletePtr) {
        return root;
      }

      const int currentVal = (*toDeletePtr)->val;

      if (currentVal == key) {
        break;
      }

      if (currentVal > key) {
        toDeletePtr = &((*toDeletePtr)->left);
      } else {
        toDeletePtr = &((*toDeletePtr)->right);
      }
    }

    // Remove the target node
    auto* toDelete = *toDeletePtr;

    if (toDelete->left && toDelete->right) {
      // Find predecessor
      auto** predPtr = &(toDelete->left);

      while ((*predPtr)->right) {
        predPtr = &((*predPtr)->right);
      }

      // Move predecessor into node to delete's position
      auto* pred = *predPtr;
      *predPtr = pred->left;

      pred->left = toDelete->left;
      pred->right = toDelete->right;

      *toDeletePtr = pred;
    } else if (toDelete->left) {
      *toDeletePtr = toDelete->left;
    } else {
      *toDeletePtr = toDelete->right;
    }

    // Free the node and return root of tree
    delete toDelete;

    return root;
  }
};
// @leet end

File: leetcode/450.delete-node-in-a-bst.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: TreeNode | None, key: int) -> TreeNode | None:
        if not root or root.val == key:
            return self.remove(root)

        prev: TreeNode | None = None
        curr = root

        while curr and curr.val != key:
            prev = curr
            curr = curr.left if key < curr.val else curr.right

        if prev.left == curr:
            prev.left = self.remove(prev.left)
        else:
            prev.right = self.remove(prev.right)

        return root

    def remove(self, node: TreeNode | None) -> TreeNode | None:
        if not node:
            return None

        if not node.left:
            return node.right

        if not node.right:
            return node.left

        prev = node
        curr = node.left

        while curr and curr.right:
            prev = curr
            curr = curr.right

        node.val = curr.val

        if prev == node:
            prev.left = curr.left
        else:
            prev.right = curr.left

        return node


# @leet end
